DM-14153 count measured and reference stars after outlier rejection

parejkoj opened the pull request.

r-owen commented on python/lsst/jointcal/jointcal.py:
```diff
@@ -668,12 +678,26 @@ def _fit_astrometry(self, dataName=None):
+    def check_stars(self):
+        """Count measured and reference stars per ccd and warn/log them."""
+        for ccdImage in self.associations.getCcdImageList():
+            stars = ccdImage.countStars()
```
I strongly suggest unpacking the returned value into `nMeasStars` and `nRefStars` or similar. I think it would make the code clearer, because the name `stars` sounds like a collection of star objects, not a count or pair of counts, and `stars` gives no hint as to which component is which.

r-owen commented on include/lsst/jointcal/CcdImage.h:
```diff
@@ -72,6 +72,13 @@ class CcdImage {
+    /**
+     * Count the number of valid measured and reference stars that fall within this ccdImage.
+     *
+     * @return (measured, reference) stars in the image.
```
Consider adding "number of" before (measured", just to emphasize that this is a pair of counts, not a pair of collections of star objects. It duplicates the brief description, so I certainly won't insist.

r-owen commented on python/lsst/jointcal/jointcal.py:
```diff
@@ -668,12 +678,26 @@ def _fit_astrometry(self, dataName=None):
+            if stars[1] < self.config.minRefStarsPerCcd:
+                self.log.warn("ccdImage %s has only %s RefStars (desired %s)",
+                              ccdImage.getName(), stars[1], self.config.minRefStarsPerCcd)
+
```
Can you do more than just log a warning if a count is too low? Is there a threshold for either value below which the fit should give up (e.g. one or two additional config parameters)?

parejkoj replied:
That is a very good question. I think I'd rather start by issuing warnings now, and turn it into a `raise` or equivalent after we've gotten a sense of how often it triggers. Right now, I just want to know about it as part of debugging things, and I don't really have a good sense for what an appropriately "safe" value is a priori. @TallJimbo might have opinions there, though?

TallJimbo replied:
It's probably possible to come up with a minimum for at least some specific models from first principles, but I think it's probably most productive to just log things for now.

r-owen commented on src/CcdImage.cc:
```diff
@@ -130,6 +130,20 @@ CcdImage::CcdImage(afw::table::SourceCatalog &catalog, std::shared_ptr<lsst::afw
+std::pair<int, int> CcdImage::countStars() const {
+    int measuredStars = 0;
+    int refStars = 0;
+    for (auto &measuredStar : _catalogForFit) {
```
please insert a `const` after `auto` since the code does not modify `_catalogForFit`

parejkoj commented:
I've reverted the "add associations to jointcal instance" commit and reworked the other changes to operate by passing `associations` around.

I wanted to add a test for `check_stars()`, but I'm not sure how to go about it: all it does is send a `log.warn` message, and I don't know how to capture that in a test to check it.

r-owen commented on include/lsst/jointcal/Associations.h:
```diff
@@ -44,12 +46,27 @@ class Associations {
+    /**
+     * Sets a shared tangent point for all ccdImages, using the mean of the centers of all ccdImages.
+     */
+    void setCommonTangentPoint();
```
Consider a name such `computeCommonTangentPoint`. It is less confusing than a `set` method that takes no arguments (usually `set` takes the value to be set). Plus less overloading makes the Python easier to understand.

r-owen commented on python/lsst/jointcal/testUtils.py:
```diff
@@ -0,0 +1,290 @@
+
+    This uses the butler dataset in `tests/data/cfht_minimal` to bootstrap the
+    metadata. Otherwise, the user has to supply the location of a butler
+    repository with the necessary metadata.
```
Either support this or remove the comment.

r-owen commented on python/lsst/jointcal/testUtils.py:
```diff
@@ -0,0 +1,290 @@
+       Result struct with components:
+
+       - `catalog` : Catalogs containing fake sources
+           (`lsst.afw.table.SourceCatalog`).
```
Please swap the type and the description here and for `ccdImage` immediately below and clean up `bbox` immediately after that. I guess it's not a `Parameters` list and maybe it doesn't matter, but it's a bit odd being different than parameter lists for no readily apparently reason (and different item by item).

parejkoj replied:
That's how Struct returns are supposed to be documented: https://developer.lsst.io/python/numpydoc.html#struct-return-types

parejkoj replied:
Oh, and the reason `bbox` looks different is because it's on the same line.

r-owen commented on python/lsst/jointcal/testUtils.py:
```diff
@@ -0,0 +1,290 @@
+    visitInfo = butler.get("calexp_visitInfo", dataId=dataId)
+    bbox = butler.get('calexp_bbox', dataId=dataId)
+    detector = butler.get('calexp_detector', dataId=dataId)
+    filt = butler.get("calexp_filter", dataId=dataId).getName()
```
At risk of being too picky, please consider being uniform in your use of single quotes or double quotes for these strings. In the 5 lines above you have 2 one way and 3 the other. It works, and "a foolish consistency" and all, so ignore if you prefer.

r-owen commented on python/lsst/jointcal/testUtils.py:
```diff
@@ -0,0 +1,290 @@
+    ----------
+    catalog : `lsst.afw.table.SourceCatalog`
+        The table to get sources from.
+    pixToFocal : `lsst.afw.geom.Transform`
```
I think you mean `lsst.afw.geom.TransformPoint2ToPoint2`

`Transform` is not actually a class in Python (though it is a templated class in C++)

r-owen commented on include/lsst/jointcal/CcdImage.h:
```diff
@@ -72,6 +72,13 @@ class CcdImage {
+    /**
+     * Count the number of valid measured and reference stars that fall within this ccdImage.
+     *
+     * @return Number of (measured, reference) stars in the image.
```
please add a few words explaining whether reference stars that are associated with an invalid measured star are counted or excluded (with some sort of justification, if practical; it may not be worth the trouble, since this count is only intended as a simple diagnostic measure and it may not be clear yet -- or ever -- which option is preferable).

r-owen commented on tests/test_ccdImage.py:
```diff
@@ -0,0 +1,102 @@
+        self.associations.setCommonTangentPoint()
+        self.associations.associateCatalogs(matchCut)
+        self.assertEqual(measStars, nStars)
+        self.assertEqual(refStars, 0)
```
How can one tell that calling setCommonTangentPoint and associateCatalogs did anything?

parejkoj replied:
If they don't raise an exception, I guess.

parejkoj replied:
An offline conversation cleared this point up: I've added a couple of sanity checks around this block to show what's happening.

timj commented on bin.src/plot_photoCalib.py:
```diff
@@ -140,7 +140,7 @@ def makePhotoCalibImages(visit, butler, step=8, chips=[], tract=None,
 
 
-class ImageMaker(object):
+class ImageMaker():
```
I'd tend to agree that it's cleaner to remove the parens.

parejkoj replied:
I didn't even know I could! I definitely agree.
